File: vscode/src/ime/input/RimeImeProvider.ts
import { execFile, execSync } from 'child_process';
import { promisify } from 'util';
import * as fs from 'fs';
import * as path from 'path';
import {
  ImeAsciiModeSwitcher,
  ImePartialState,
  ImeProvider,
  ImeStateSource,
  ImeType,
  Logger,
} from '../../core/types';
import { nativeRimeInputState, nativeWaitForRimeStateChange } from '../system/native';

const execFileAsync = promisify(execFile);
const SEARCH_BASE_DIRS = [
  'C:\\Program Files (x86)\\Rime',
  'C:\\Program Files\\Rime',
  'D:\\Program Files\\Rime',
];

function detectWeaselServer(logger?: Logger): string | null {
  const fromRegistry = readRegistry(logger);
  return fromRegistry ?? scanDirectories(logger);
}

function readRegistry(logger?: Logger): string | null {
  try {
    const output = execSync(
      'reg query "HKLM\\SOFTWARE\\Rime\\Weasel" /v WeaselRoot',
      { encoding: 'utf-8', timeout: 5000 },
    );
    const match = output.match(/WeaselRoot\s+REG_SZ\s+(.+)/);
    if (match?.[1]) {
      const serverPath = path.join(match[1].trim(), 'WeaselServer.exe');
      if (fs.existsSync(serverPath)) {
        logger?.info(`找到 WeaselServer (注册表): ${serverPath}`);
        return serverPath;
      }
    }
  } catch {
  }
  return null;
}

function scanDirectories(logger?: Logger): string | null {
  const candidates: { version: string; path: string }[] = [];
  for (const basePath of SEARCH_BASE_DIRS) {
    if (!fs.existsSync(basePath)) continue;
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(basePath, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const entry of entries) {
      if (!entry.isDirectory() || !entry.name.startsWith('weasel-')) continue;
      const version = entry.name.substring('weasel-'.length);
      const serverPath = path.join(basePath, entry.name, 'WeaselServer.exe');
      if (fs.existsSync(serverPath)) candidates.push({ version, path: serverPath });
    }
  }
  candidates.sort((a, b) => compareVersions(b.version, a.version));
  const result = candidates[0]?.path ?? null;
  if (result) logger?.info(`找到 WeaselServer (扫描): ${result}`);
  return result;
}

function compareVersions(a: string, b: string): number {
  const pa = a.split('.').map(Number);
  const pb = b.split('.').map(Number);
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const difference = (pa[i] ?? 0) - (pb[i] ?? 0);
    if (difference !== 0) return difference;
  }
  return 0;
}

/** Rime 通过 Lua 共享内存提供中英文/输入中状态，并使用 Weasel 指令切换中英文。 */
export class RimeImeProvider implements ImeProvider, ImeAsciiModeSwitcher, ImeStateSource {
  readonly type = ImeType.RIME;
  readonly name = 'Rime/Weasel';
  readonly asciiModeSwitcher = this;
  readonly stateSource = this;

  private weaselServerPath: string | null;
  private warnedMissing = false;
  private stateAvailable = false;

  constructor(
    private logger: Logger,
    weaselPath?: string,
  ) {
    this.weaselServerPath = weaselPath ?? detectWeaselServer(logger);
    this.logger.info(`WeaselServer path: ${this.weaselServerPath ?? '(not found)'}`);
  }

  start(): void {
  }

  readState(): ImePartialState {
    const actual = nativeRimeInputState();
    this.stateAvailable = actual != null;
    return actual == null ? {} : {
      isAsciiMode: actual.isAsciiMode,
      isComposing: actual.isComposing,
    };
  }

  isAvailable(): boolean {
    return this.stateAvailable;
  }

  supportsChangeNotifications(): boolean {
    return true;
  }

  async waitForStateChange(timeoutMillis: number): Promise<boolean> {
    const result = await nativeWaitForRimeStateChange(timeoutMillis);
    if (result < 0) {
      await new Promise((resolve) => setTimeout(resolve, timeoutMillis));
    }
    return result > 0;
  }

  async switchAsciiMode(
    ascii: boolean,
    shouldContinue: () => boolean = () => true,
  ): Promise<boolean> {
    return this.switchImeMode(
      ascii ? '/ascii' : '/nascii',
      shouldContinue,
    );
  }

  dispose(): void {
  }

  private async switchImeMode(
    arg: string,
    shouldContinue: () => boolean,
  ): Promise<boolean> {
    const exePath = this.weaselServerPath;
    if (!exePath) {
      if (!this.warnedMissing) {
        this.warnedMissing = true;
        const msg = 'WeaselServer.exe 未找到，请检查 autoSwitchIME.weaselServerPath 设置';
        this.logger.warn(msg);
      }
      return false;
    }
    if (!fs.existsSync(exePath)) {
      this.logger.warn(`WeaselServer.exe 不存在: ${exePath}`);
      return false;
    }
    try {
      if (!shouldContinue()) return false;
      await execFileAsync(exePath, [arg], { timeout: 3000, windowsHide: true });
      this.logger.info(`IME 切换: ${arg}`);
      return true;
    } catch (error) {
      this.logger.warn(`WeaselServer.exe 调用失败: ${arg}`, error as Error);
      return false;
    }
  }
}

File: core/src/main/kotlin/com/auto_switch_ime/core/ime/input/RimeImeProvider.kt
package com.auto_switch_ime.core.ime.input

import com.auto_switch_ime.core.ImeAsciiModeSwitcher
import com.auto_switch_ime.core.ImeConfig
import com.auto_switch_ime.core.ImeProvider
import com.auto_switch_ime.core.ImePartialState
import com.auto_switch_ime.core.ImeStateSource
import com.auto_switch_ime.core.ImeType
import com.auto_switch_ime.core.ime.system.NativeImeSys
import com.auto_switch_ime.core.util.Logger
import java.io.File
import java.util.concurrent.TimeUnit

/** Rime 通过 Lua 共享内存提供中英文/输入中状态，并使用 Weasel 指令切换中英文。 */
class RimeImeProvider(
    config: ImeConfig,
    private val logger: Logger
) : ImeProvider, ImeAsciiModeSwitcher, ImeStateSource {
    override val type: ImeType = ImeType.RIME
    override val name: String = "Rime/Weasel"
    override val asciiModeSwitcher: ImeAsciiModeSwitcher = this
    override val stateSource: ImeStateSource = this

    private val weaselServerPath = config.weaselServerPath ?: WeaselPathDetector.detect(logger)

    @Volatile
    private var stateAvailable = false

    override fun start() = Unit

    override fun readState(): ImePartialState {
        val actual = NativeImeSys.imeGetRimeState()
        stateAvailable = actual != null
        return actual?.let {
            ImePartialState(isAsciiMode = it.isAsciiMode, isComposing = it.isComposing)
        } ?: ImePartialState()
    }

    override fun isAvailable(): Boolean = stateAvailable

    override fun supportsChangeNotifications(): Boolean = true

    override fun waitForStateChange(timeoutMillis: Int): Boolean {
        val result = NativeImeSys.imeWaitForRimeStateChange(timeoutMillis)
        if (result < 0) {
            try {
                Thread.sleep(timeoutMillis.toLong())
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        }
        return result > 0
    }

    override suspend fun switchAsciiMode(
        ascii: Boolean,
        shouldContinue: () -> Boolean
    ): Boolean {
        val path = weaselServerPath
        if (path == null) {
            logger.warn("WeaselServer.exe not found")
            return false
        }
        if (!File(path).exists()) {
            logger.warn("WeaselServer.exe not exists at: $path")
            return false
        }

        val argument = if (ascii) "/ascii" else "/nascii"
        val label = if (ascii) "ASCII" else "Chinese"
        return try {
            if (!shouldContinue()) return false
            logger.debug("Executing: $path $argument")
            val process = ProcessBuilder(path, argument)
                .redirectOutput(ProcessBuilder.Redirect.DISCARD)
                .redirectError(ProcessBuilder.Redirect.DISCARD)
                .start()
            val exited = process.waitFor(1, TimeUnit.SECONDS)
            if (exited) {
                logger.info("Switched to $label mode (exitCode=${process.exitValue()})")
                process.exitValue() == 0
            } else {
                logger.warn("Switch to $label mode timed out")
                process.destroy()
                false
            }
        } catch (e: Exception) {
            logger.warn("Failed to switch IME mode: $argument", e)
            false
        }
    }

    override fun dispose() = Unit
}
